limitador/src/storage/mod.rs

```diff
@@ -21,38 +23,230 @@ pub enum Authorization<'c> {
     Limited(&'c Counter), // First counter found over the limits
 }
 
-pub trait Storage: Sync + Send {
-    fn get_namespaces(&self) -> Result<HashSet<Namespace>, StorageErr>;
-    fn add_limit(&self, limit: &Limit) -> Result<(), StorageErr>;
-    fn get_limits(&self, namespace: &Namespace) -> Result<HashSet<Limit>, StorageErr>;
-    fn delete_limit(&self, limit: &Limit) -> Result<(), StorageErr>;
-    fn delete_limits(&self, namespace: &Namespace) -> Result<(), StorageErr>;
+pub struct Storage {
+    limits: RwLock<HashMap<Namespace, HashSet<Limit>>>,
+    counters: Box<dyn CounterStorage>,
+}
+
+pub struct AsyncStorage {
+    limits: RwLock<HashMap<Namespace, HashSet<Limit>>>,
+    counters: Box<dyn AsyncCounterStorage>,
+}
+
+impl Storage {
+    pub fn new() -> Self {
+        Self {
+            limits: RwLock::new(HashMap::new()),
+            counters: Box::new(InMemoryStorage::default()),
+        }
+    }
+
+    pub fn with_counter_storage(counters: Box<dyn CounterStorage>) -> Self {
+        Self {
+            limits: RwLock::new(HashMap::new()),
+            counters,
+        }
+    }
+
+    pub fn get_namespaces(&self) -> HashSet<Namespace> {
+        self.limits.read().unwrap().keys().cloned().collect()
+    }
+
+    pub fn add_limit(&self, limit: Limit) {
+        let namespace = limit.namespace().clone();
+        self.limits
+            .write()
+            .unwrap()
+            .entry(namespace)
+            .or_default()
+            .insert(limit);
+    }
+
+    pub fn get_limits(&self, namespace: &Namespace) -> HashSet<Limit> {
+        match self.limits.read().unwrap().get(namespace) {
+            Some(limits) => limits.iter().cloned().collect(),
+            None => HashSet::new(),
+        }
+    }
+
+    pub fn delete_limit(&self, limit: &Limit) -> Result<(), StorageErr> {
+        let mut limits = HashSet::new();
+        limits.insert(limit.clone());
+        self.counters.delete_counters(limits)?;
+
+        let mut limits = self.limits.write().unwrap();
+
+        if let Some(counters_by_limit) = limits.get_mut(limit.namespace()) {
```

Comment:
`counters_by_limit` Not sure about what that name means or if it's correct?

Reply:
No it's not... left over from the code I moved… I'll fix! Good catch thanks!